packages/zen-bull-vault/src/EmergencyWithdraw.sol

```diff
@@ -0,0 +1,131 @@
+// SPDX-License-Identifier: GPL-3.0-only
+pragma solidity =0.7.6;
+pragma abicoder v2;
+
+// interface
+import { IERC20 } from "openzeppelin/token/ERC20/IERC20.sol";
+import { IZenBullStrategy } from "./interface/IZenBullStrategy.sol";
+import { IWETH9 } from "squeeth-monorepo/interfaces/IWETH9.sol";
+// contract
+import { ERC20 } from "openzeppelin/token/ERC20/ERC20.sol";
+import { UniFlash } from "./UniFlash.sol";
+// lib
+import { StrategyMath } from "squeeth-monorepo/strategy/base/StrategyMath.sol";
+import { Address } from "openzeppelin/utils/Address.sol";
+
+contract EmergencyWithdraw is ERC20, UniFlash {
+    using StrategyMath for uint256;
+    using Address for address payable;
+
+    /// @dev ZenBull total supply amount at the time of this contract deployment
+    uint256 private _zenBullSupply;
+
+    /// @dev crab strategy address
+    address internal immutable crab;
+    /// @dev ZenBull strategy address
+    address internal immutable zenBull;
+    /// @dev WETH address
+    address internal immutable weth;
+    /// @dev wPowerPerp address
+    address internal immutable wPowerPerp;
+
+    event Withdraw(
+        address indexed recepient,
+        uint256 zenBullAmountRedeemed,
+        uint256 crabAmountRedeemed,
+        uint256 wPowerPerpRedeemed,
+        uint256 ethReceived,
+        uint256 eulerRecoveryTokenAmount
+    );
+
+    /**
+     * @dev constructor
+     * @param _crab crab address
+     * @param _zenBull ZenBull address
+     * @param _weth WETH address
+     * @param _wPowerPerp WPowerPerp address
+     * @param _factory Uni V3 factory contract
+     */
+    constructor(
+        address _crab,
+        address _zenBull,
+        address _weth,
+        address _wPowerPerp,
+        address _factory
+    ) ERC20("ZenBullEulerRecovery", "ZBER") UniFlash(_factory) {
+        crab = _crab;
+        zenBull = _zenBull;
+        weth = _weth;
+        wPowerPerp = _wPowerPerp;
+
+        _zenBullSupply = IERC20(_zenBull).totalSupply();
+
+        IERC20(_wPowerPerp).approve(_zenBull, type(uint256).max);
+    }
+
+    /**
+     * @dev receive ETH
+     */
+    receive() external payable {
+        require(msg.sender == weth, "Can't receive ETH from this sender");
+    }
+
+    /**
+     * @notice withdraw ETH deposited into crab
+     * @dev this will give the sender ZBEPR token as ownership for the ETH deposited in Euler pool
+     * @param _zenBullAmount ZenBull amount to redeem
+     * @param _maxEthForWPowerPerp max ETH to pay for flashswapped oSQTH amount
+     */
+    function emergencyWithdrawEthFromCrab(uint256 _zenBullAmount, uint256 _maxEthForWPowerPerp)
```

Comment:
A thought: It might be nice to have an alternative withdraw where they provide oSQTH rather than needing to trade through the pool. 

They could achieve the sam thing by putting their oSQTH in a 1-tick LP and calling this, but more complicated and worse gas.